### PR TITLE
Fix duplicate detection in #add_to_index.

### PR DESCRIPTION
**Pull request by bartkamphorst**
This fixes https://github.com/gollum/gollum/issues/1416.

### CODE REVIEW

File: lib/gollum-lib/committer.rb [hunk 1]
```diff
@@ -75,14 +75,17 @@ def parents
...
     # This way, pages are not inadvertently overwritten.
     #
     # Returns nothing (modifies the Index in place).
-    def add_to_index(path, data, options = {})
+    def add_to_index(path, data, options = {}, force_overwrite = false)
       if index.current_tree
-        tree = @wiki.page_file_dir ? index.current_tree / @wiki.page_file_dir : index.current_tree
-        raise DuplicatePageError.new(path) if !page_path_scheduled_for_deletion?(index.tree, path) && tree.blobs.find {|blob| blob.name == path}
+        tree = index.current_tree
+        unless force_overwrite
```
Comment (dometto) [inline lib/gollum-lib/committer.rb]: Maybe: `unless force_overwrite || page_path_scheduled_for_deletion?(index.tree, path)`

If it's a delete operation, there is no need to call `tree / path`.

File: lib/gollum-lib/committer.rb [hunk 2]
```diff
@@ -75,14 +75,17 @@ def parents
...
+          tree = tree / path
+          raise DuplicatePageError.new(path) unless tree.nil? || page_path_scheduled_for_deletion?(index.tree, path)
```
Comment (dometto) [inline lib/gollum-lib/committer.rb]: If you adopts the above suggestion, remove the second unless-disjunct here

File: lib/gollum-lib/committer.rb [hunk 3]
```diff
@@ -75,14 +75,17 @@ def parents
...
+    # Raises Gollum::DuplicatePageError if a matching filename already exists, unless force_overwrite is explicitly enabled.
     # This way, pages are not inadvertently overwritten.
     #
     # Returns nothing (modifies the Index in place).
-    def add_to_index(path, data, options = {})
+    def add_to_index(path, data, options = {}, force_overwrite = false)
       if index.current_tree
-        tree = @wiki.page_file_dir ? index.current_tree / @wiki.page_file_dir : index.current_tree
-        raise DuplicatePageError.new(path) if !page_path_scheduled_for_deletion?(index.tree, path) && tree.blobs.find {|blob| blob.name == path}
+        tree = index.current_tree
```
Comment (dometto) [inline lib/gollum-lib/committer.rb]: Wait, should this not respect `page_file_dir`? (Compare the deleted line 84)
```ruby
tree = index.current_tree
tree = tree / @wiki.page_file_dir if @wiki.page_file_dir
```

File: test/test_wiki.rb [hunk 4]
```diff
@@ -303,18 +303,41 @@
...
+      @wiki.write_file("Abc-Def.file", "# Gollum", commit_details)
+    end
+    @wiki.write_file("subdir/Abc-Def.file", "# Gollum", commit_details)
+    assert_raises Gollum::DuplicatePageError do
+      @wiki.write_file("subdir/Abc-Def.file", "# Gollum", commit_details)
+    end
+  end
+
+  test "write page can overwrite pages if overwrite option is enabled" do
+    omit('Not yet implemented.')
```
Comment (dometto) [inline test/test_wiki.rb]: Ok to omit this for now, but for my comprehension: is the plant to implement this in a later release, or a separate PR?

File: test/test_wiki.rb [hunk 5]
```diff
@@ -303,18 +303,41 @@
     assert @wiki.page("Saruman")
   end
 
-  test "write page is not allowed to overwrite file" do
+  test "write page is not allowed to overwrite pages by default" do
```
Comment (dometto) [inline test/test_wiki.rb]: Is there a possibility to run these tests against a wiki with `page_file_dir` set, just to test whether it works with that?

**Comment by dometto** [comment]
Close/reopening to fix tests

File: lib/gollum-lib/committer.rb (hunk 1 above)
Comment (bartkamphorst) [reply lib/gollum-lib/committer.rb]: Good suggestion, will implement.

File: lib/gollum-lib/committer.rb (hunk 3 above)
Comment (bartkamphorst) [reply lib/gollum-lib/committer.rb]: Both `index.current_tree` and `path` contain `page_file_dir` if it is set, so `tree = tree / path` works either way.

File: test/test_wiki.rb (hunk 4 above)
Comment (bartkamphorst) [reply test/test_wiki.rb]: In a separate PR, and likely in a later release as well. Will actually remove these tests from this PR.

**Pull request merged**